jpyinterpreter-docs/src/modules/ROOT/pages/jpyinterpreter-architecture/jpyinterpreter-architecture.adoc

```diff
@@ -0,0 +1,125 @@
+
+** Setup cells variables.
+
+** Delegating to `PythonGeneratorTranslator` when it detects the function being translated is a generator.
+
+** Using `FlowGraph` to calculate the `StackMetadata` of each instruction.
+
+** Calling the `implement` method on every opcode in the `Opcode` list with the `StackMetadata` for the opcode and the `FunctionMetadata` for the overall function.
+
+- `PythonGeneratorTranslator` is like `PythonBytecodeToJavaBytecodeTranslator` but for generators.
+It breaks a single generator function into multiple `advance` functions, and generate each `advance` function bytecode independently.
+
+- `FlowGraph`, which calculates the `StackMetadata` that corresponds to each `Opcode`.
+It is responsible for unifying the `StackMetadata` from all jump sources for each `Opcode` that is a jump target.
+For instance, if two sources with the same target have post `StackMetadata` of `... int` and `... bool` respectively, `FlowGraph` will unify that to `... int` (since `bool` is a subclass of `int`, for better or worse).
+
+- `StackMetadata` stores metadata about the stack and local variables.
+Each `Opcode` get its own `StackMetadata` instance.
+It is mostly used to perform optimizations; for instance, if we detect the top two items on the stack are `int` and `int` for the `BINARY_ADD` instruction, we can change the (normally complex due to Python semantics) `BINARY_ADD` bytecode into a single method call.
+
+- `FunctionMetadata` stores metadata about the function (for instance, the `MethodVisitor` to use to generate bytecode). Each `Opcode` get the same `FunctionMetadata` instance.
+
+- `Opcode` are the interface between CPython opcodes and the `Implementors`.
+Each describe a particular operation, and are usually (but not always) correspond to a CPython opcode.
+Some CPython opcodes will map to the same `Opcode` implementation.
+
+- `Implementors` are responsible for generating the Java bytecode corresponding to CPython bytecode.
+They can be found in the `implementors` package.
+
+In a diagram, the overall process of compiling a function looks like this:
+
+image::jpyinterpreter-architecture/jpyinterpreter-architecture.png[A diagram showing how JPyInterpreter classes interact]
+
+
+== Types
+
+The builtin types for JPyInterpreter can be found in the `types` package.
+They all implement `PythonLikeObject`, which the interface the bytecode uses to represent arbitrary objects.
+If type flow analysis determines a more specific type can be used (via `StackMetadata`), the more specific type is used directly instead.
+`PythonLikeObject` have several methods:
+
+- `\\__getAttributeOrNull`: returns the attribute with the given name if it exists, otherwise returns null.
+This is NOT `__getattribute\__` (which is implemented by `$method$\__getattribute\__` instead).
+This is more akin to `self.\__dict\__[attribute]`.
+The default `$method$\__getattribute__` uses it to get the attribute (with additional magic to handle descriptors).
```

Comment:
Maybe there should be a place to explain the magic? You don't usually see the word "magic" in developer documentation. :-)